Add an explicit enum value to the spec for text and icon pitch and rotation

boundsj opened the pull request:
This implements the solution style spec updates described in https://github.com/mapbox/mapbox-gl-native/pull/6253#issuecomment-244856555

> one possible approach to address the special-caseness of these properties would be to add an explicit enum value to the spec like auto or inherit for text-rotation-alignment, icon-rotation-alignment and text-pitch-alignment.
> - Add auto enum value
> - Set default value to auto

cc @lbud @yhahn @jfirebaugh @1ec5

yhahn committed:
Sketch of removing spec-defined default value for text-rotation-alignment given that its renderer-defined defaults are more complex.

yhahn committed:
Unset default for icon-rotation-alignment as well.

boundsj committed:
Introduce new default enum value for text and icon rotation and pitch

This adds a new enumeration value called `auto` for
text-rotation-alignment, icon-rotation-alignment and
text-pitch-alignment. The new value is set as the default for the
three properties it was added to and the associated `doc`
description was updated to reflect that.

In addition, the relevant SDK support tables have been updated to
reflect the first expected native SDK releases that'll support the new
`auto` value. The js SDK version was not updated since the value was
not known at the time of this commit. It can be updated in the future.

jfirebaugh commented:
Rather than change the value of the "basic" level of "sdk-support", I think we should add a separate key "'auto' value". And support [here](https://github.com/mapbox/mapbox-gl-style-spec/blob/885e2bdcaa715915bb7b97132008db9a6c90e950/docs/_generate/item.html#L85-L86) for rendering arbitrary keys from "sdk-support".

1ec5 commented on reference/v8.json:
```diff
@@ -790,15 +791,19 @@
       ],
-      "doc": "Aligns text to the plane of the `viewport` or the `map` when the map is pitched. Matches `text-rotation-alignment` if unspecified.",
+      "default": "auto",
+      "doc": "Aligns text to the plane of the `viewport` or the `map` when the map is pitched. Matches `text-rotation-alignment` if auto.",
```
Nit: surround `auto` in backticks. (That way it’ll become `MGLTextPitchAlignmentAuto` when the iOS/macOS SDK documentation is automatically generated from this file.)

boundsj committed:
Generate sdk support rows from spec JSON

This adds logic to loop through all of the values in `sdk-support`
and render the appropriate documentation for each of them. There
are special cases for `basic` -> "Base" and to ignore
`property-function` since that is already handled as a special
case and has an associated bool value at the root level of each
item's object in the spec.

boundsj committed:
Surround auto in backticks

This will cause it to become MGLTextPitchAlignmentAuto when the
iOS/macOS SDK documentation is automatically generated from this file.

boundsj commented:
>  I think we should add a separate key "'auto' value". And support here for rendering arbitrary keys from "sdk-support".

This is done in https://github.com/mapbox/mapbox-gl-style-spec/pull/493/commits/4b8bfbd0301a822de57934085433dfbb401b1229. I used the expected native platform versions that `auto` will be supported in. I stubbed `js` as `TBD` since I'm not certain how the this change and the next js release that supports it will be orchestrated.

Also, the `auto` string where present in the `doc` value gained backticks for the benefit of ios/macos documentation

cc @jfirebaugh @lucaswoj @lbud @1ec5

1ec5 commented on reference/v8.json:
```diff
@@ -790,15 +796,24 @@
       ],
-      "doc": "Aligns text to the plane of the `viewport` or the `map` when the map is pitched. Matches `text-rotation-alignment` if unspecified.",
+      "default": "auto",
+      "doc": "Aligns text to the plane of the `viewport` or the `map` when the map is pitched. Matches `text-rotation-alignment` if `auto`.",
```
While you're here, I think the use of backticks here would contribute to confusing documentation in the iOS and macOS SDKs. It'll say something like "Aligns text to the plane of the `MGLTextPitchAlignmentViewport` or the `MGLTextPitchAlignmentMap` when the map is pitched." But the intent is to say "plane of the viewport when set to `viewport`".

boundsj replied:
Yeah I see what you mean. In 5ab221e I updated so it now reads:

#### spec docs

<img width="632" alt="screen shot 2016-09-15 at 9 14 40 pm" src="https://cloud.githubusercontent.com/assets/63186/18575212/e1e2b390-7b89-11e6-8bb5-aa6e5d9f04c6.png">

#### iOS SDK docs

<img width="1254" alt="screen shot 2016-09-15 at 9 14 27 pm" src="https://cloud.githubusercontent.com/assets/63186/18575213/e1e46bd6-7b89-11e6-901d-01ff69f59a19.png">

boundsj committed:
Clarify documentation for text-pitch-alignment

lucaswoj commented on reference/v8.json:
```diff
@@ -607,6 +608,11 @@
           "android": "2.0.1"
+        },
+        "auto": {
+          "js": "TBD",
```
I assume "TBD" stands for "to be decided." Why "TBD" instead of leaving this `undefined` to indicate that the feature is not supported?

boundsj replied:
Yeah that was my question in https://github.com/mapbox/mapbox-gl-style-spec/pull/493#issuecomment-247483816. Sounds like the process is to land this PR first and leave all 3 SDKs as `undefined`. Then, land the SDK PRs to support `auto`, then come back here and make a new PR to update support. Is my understanding correct?

boundsj replied:
@lucaswoj 296d335 makes support for all SDKs for undefined for `auto` for now. I'm guessing about how a style spec change this should all be orchestrated to the SDKs so please let me know when we should add the specific version (at least on the GL JS side). 

https://github.com/mapbox/mapbox-gl-js/pull/3203 has added support for `auto` to GL JS but still needs review and the tests in https://github.com/mapbox/mapbox-gl-test-suite/pull/144

lucaswoj replied:
> let me know when we should add the specific version (at least on the GL JS side)

Our policy is to update the style spec once the feature goes out in a public release. The release coordinator is responsible for this update.

lucaswoj commented on reference/v8.json:
```diff
@@ -595,10 +595,11 @@
       ],
-      "default": "viewport",
-      "doc": "Orientation of icon when map is rotated.",
+      "default": "auto",
```
Where do we explain what `auto` means?

boundsj replied:
@lucaswoj do the changes to the doc strings in 296d335 address your concerns about the lack of explanation of `auto`?

lucaswoj replied:
Looks good! Thanks 😄

boundsj committed:
Rebuild minified file

boundsj committed:
Remove sdk support and enhance doc description for auto

This updates the doc descriptions for text-rotation-alignment
and icon-rotation-alignment so that auto has a meaning in the
system (as a default).

This also removes all SDK support for auto. Once support for auto
lands in GL JS and GL native then the specific version numbers
can be added back.

jfirebaugh reviewed (changes requested):
I want to suggest a slightly different way of doing the SDK support matrix -- will push to a separate branch for discussion.

jfirebaugh committed:
Generate markdown with remark

jfirebaugh committed:
Allow custom SDK support matrix headers to use markdown

jfirebaugh commented:
https://github.com/mapbox/mapbox-gl-style-spec/compare/symbol-complex-default...symbol-complex-default-alt
- Supports markdown in 1st column
- Custom rows appear after "Base support" and "Data-driven styling"

![image](https://cloud.githubusercontent.com/assets/98601/18643341/8db019fc-7e59-11e6-9cfa-b56c49b8b882.png)

If you are 👍  please pull into this branch and merge.

lucaswoj commented:
Why "Base _support_" but not "Data-driven styling _support_" or "`auto` value _support_"?

jfirebaugh commented:
@lucaswoj This PR doesn't change that; let's not nitpick it here.

boundsj commented:
@jfirebaugh 👍 I merged the commits from your branch
